File: 51.py
class Solution:
    # @return a list of lists of string
    def solveNQueens(self, n):
        total_sol = []
        sol = []
        i = 1
        j = 1
        while i > 0:
            while j < n+1:
                if self.checkValidity(i,j,sol):
                    sol.append([i,j])
                    if i == n:
                        total_sol.append(sol[:])
                        sol.pop()
                        break
                    else:
                        i += 1                      
                        j = 1
                else:
                    j += 1       
            i -= 1
            if i <= 0:
                break
            j = sol[-1][1] + 1
            sol.pop()
        
        transfer_total_sol = self.transfer(total_sol,n)
        return transfer_total_sol
    
    def transfer(self,total_sol, n):
        transfer_total_sol = []
        for sol in total_sol:
            baseform = [ ['.' for x in range(n)] for y in range(n) ]
            for pair in sol:
                i = pair[0]
                j = pair[1]
                baseform[i-1][j-1] = 'Q'
            final_str = ''
            for k in range(n):
                baseform[k] = ''.join(baseform[k])
            transfer_total_sol.append(baseform[:])
             
        return transfer_total_sol
    def checkValidity(self,row,col,sol):
        for pair in sol:
            r = pair[0]
            c = pair[1]
            if row == r or col == c:
                return False
            if abs(row-r) == abs(col-c):
                return False
        return True
        
    # new solution #####################
    def solveNQueens(self, n):
        stack, res = [[(0, i)] for i in range(n)], []
        while stack:
            board = stack.pop()
            row = len(board)
            if row == n:
                res.append([''.join('Q' if i == c else '.' for i in range(n))
                            for r, c in board])
            for col in range(n):
                if all(col != c and abs(row-r) != abs(col-c)for r, c in board):
                    stack.append(board+[(row, col)])
        return res


so = Solution()       
print so.solveNQueens(3)    

File: 51.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

class Solution {
public:
    vector<vector<string> > result ;
    vector<vector<string> > solveNQueens(int n){
        vector<int> board(n,0) ;
        solve(board,0,n);
        return result;
    }


    void solve(vector<int> board, int row,int n){

        if (row == n){
            vector<string> str_board = transfer(board);
            result.push_back(str_board);
            return;
        }
        for(int col = 0 ; col < n; col++){
            if(isValid(board,row,col)){
                board[row] = col;
                solve(board,row+1,n);
            }
        }

        return;
    }



    vector<string>  transfer(vector<int> board){
        int n = board.size();
        vector<string> str_board;
        int col = 0;
        for(int row =0; row<n;row++)
        {
            string s(n,'.');
            col = board[row];
            s[col] = 'Q';
            str_board.push_back(s);
        }

        return str_board;
    }


    bool isValid(vector<int> board,int row,int col){

        for(int r = row-1; r>=0 ;r--)
        {
            if (board[r] == col || abs(r-row) == abs(board[r] - col) )
                return false;
        }
        return true;

    }

};


int main(){

    Solution sol;
    vector<vector<string> > result = sol.solveNQueens(4);
    int m = result.size();
    int n = result[0].size();
    for(int i =0; i<m;i++)
        for(int j =0 ; j<n;j++)
            cout << result[i][j]<<endl;
    return 0;
}